File: 02-development/src/backend/app/sla.py
from datetime import datetime, timezone

from app.schemas import Priority, Status, WaitlistEntry

ACTIVE_STATUSES = (Status.waiting, Status.notified)

# Priority rank per docs/spec.md 3.3: vip > reservation_overflow > standard.
_PRIORITY_RANK = {
    Priority.vip: 0,
    Priority.reservation_overflow: 1,
    Priority.standard: 2,
}


def is_sla_breached(entry: WaitlistEntry, now: datetime | None = None) -> bool:
    """docs/spec.md 3.4: breached when now - created_at > quoted_wait_minutes,
    only meaningful while the entry is still active in the queue."""
    if entry.status not in ACTIVE_STATUSES:
        return False
    now = now or datetime.now(timezone.utc)
    elapsed_minutes = (now - entry.created_at).total_seconds() / 60
    return elapsed_minutes > entry.quoted_wait_minutes


def _closed_at(entry: WaitlistEntry) -> datetime:
    return entry.seated_at or entry.cancelled_at or entry.created_at


def queue_sort_key(entry: WaitlistEntry) -> tuple:
    """Sort key implementing compareByQueueOrder from
    src/frontend/src/utils/sla.ts: active entries first (by priority rank,
    then created_at ascending), closed entries sink below (most-recently-
    closed first)."""
    is_active = entry.status in ACTIVE_STATUSES
    if is_active:
        return (0, _PRIORITY_RANK[entry.priority], entry.created_at)
    # Closed entries: reverse-chronological (most recent first) within the
    # closed group, which sorts after all active entries.
    return (1, -_closed_at(entry).timestamp(), entry.created_at)

File: 02-development/src/frontend/src/utils/sla.ts
import type { Status, WaitlistEntry } from "../types/waitlist";

const ACTIVE_STATUSES: Status[] = ["waiting", "notified"];

// SLA breach per docs/spec.md 3.4: now - created_at > quoted_wait_minutes,
// only meaningful while the entry is still active in the queue.
export function isSlaBreached(entry: WaitlistEntry, now: Date = new Date()): boolean {
  if (!ACTIVE_STATUSES.includes(entry.status)) return false;
  const elapsedMinutes = (now.getTime() - new Date(entry.created_at).getTime()) / 60000;
  return elapsedMinutes > entry.quoted_wait_minutes;
}

export function elapsedMinutes(createdAt: string, now: Date = new Date()): number {
  return Math.max(0, Math.floor((now.getTime() - new Date(createdAt).getTime()) / 60000));
}

// Priority rank per docs/spec.md 3.3: vip > reservation_overflow > standard.
const PRIORITY_RANK = { vip: 0, reservation_overflow: 1, standard: 2 } as const;

function closedAt(entry: WaitlistEntry): number {
  return new Date(entry.seated_at ?? entry.cancelled_at ?? entry.created_at).getTime();
}

// Section 3.3's priority/created_at ordering describes the *active* queue.
// Closed-out entries (seated/cancelled/no_show) aren't waiting for anything
// anymore, so they sink below every active entry regardless of priority,
// most-recently-closed first.
export function compareByQueueOrder(a: WaitlistEntry, b: WaitlistEntry): number {
  const aActive = ACTIVE_STATUSES.includes(a.status);
  const bActive = ACTIVE_STATUSES.includes(b.status);
  if (aActive !== bActive) return aActive ? -1 : 1;

  if (!aActive) return closedAt(b) - closedAt(a);

  const rankDiff = PRIORITY_RANK[a.priority] - PRIORITY_RANK[b.priority];
  if (rankDiff !== 0) return rankDiff;
  return new Date(a.created_at).getTime() - new Date(b.created_at).getTime();
}
